qemu/tests/rv_build_install.py

```diff
@@ -1,263 +1,112 @@
-        pkgsRequired = ["libogg-devel", "celt051-devel", "libcacard-devel",
-                        "source-highlight", "gtk-doc", "libepoxy-devel"]
-    else:
-        pkgsRequired = ["libogg-devel", "celt051-devel", "libcacard-devel"]
-
-    install_req_pkgs(pkgsRequired, vm_root_session, params)
-
-    try:
-        cmd = "yum --disablerepo=\"*\" " + \
-              "--enablerepo=\"epel\" -y install perl-Text-CSV"
-        output = vm_root_session.cmd(cmd, timeout=300)
-        logging.info(output)
-    except:
-        logging.error(output)
-
-    # spice-gtk needs to built from tarball before building virt-viewer on RHEL6
-    pkgName = params.get("build_install_pkg")
-    if pkgName != "spice-gtk":
-        tarballLocation = "http://www.spice-space.org/download/gtk/spice-gtk-0.30.tar.bz2"
-        cmd = "%s -p spice-gtk --tarball %s" % (vm_script_path, tarballLocation)
-        output = vm_root_session.cmd(cmd, timeout=600)
-        logging.info(output)
-        if re.search("Return code", output):
-            raise error.TestFail("spice-gtk was not installed properly")
+    # pkg_install()
+
+    def install(self):
+        self.setup()
+        self.pkg_install()
+        self.finish()
+    # run()
+
+    def finish(self):
+        utils_spice.clear_interface(self.vm)
+    # finish()
+
+    @staticmethod
+    def factory(name, params, env):
+        _type = None
+
+        if name == "spice-gtk3":
+            _type = BuildInstallSpiceGtk
+        elif name == "spice-vdagent":
+            _type = BuildInstallSpiceVdagent
+        elif name == "virt-viewer":
+            _type = BuildInstallVirtViewer
+        elif name == "xorg-x11-drv-qxl":
+            _type = BuildInstallQxl
         else:
-            logging.info("spice-gtk was installed")
+            raise error.TestFail("Invalid package name: '%s'" % name)
 
-    else:
-        output = vm_root_session.cmd("%s -p spice-gtk" % (vm_script_path),
-                                     timeout=600)
-        logging.info(output)
-        if re.search("Return code", output):
-            raise error.TestFail("spice-gtk was not installed properly")
+        return _type(name, params, env)
+    #factory()
+# BuildInstallBase
 
-    # Get version of spice-gtk after install
-    try:
-        output = vm_root_session.cmd("LD_LIBRARY_PATH=/usr/local/lib"
-                                     " remote-viewer --spice-gtk-version")
-        logging.info(output)
-    except:
-        logging.error(output)
 
+class BuildInstallSpiceGtk(BuildInstall):
+    def setup(self):
+        BuildInstall.setup(self)
 
-def build_install_vdagent(vm_root_session, vm_script_path, params):
-    """
-    Build and install spice-vdagent in the VM
+        # requires removal of conflicting packages in RHEL7.3
+        logging.info("Removing gnome-boxes and vinagre due to conflicts with newer spice-gtk3")
+        self.vm_root_session.cmd("yum -y remove gnome-boxes vinagre")
```

Comment:
Instead of hardcoding the packages that are conflicting at this time and then updating the code when those are not a problem anymore does not seem the right thing to do. I should probably add more keys to the configuration file, to identify such cases.